File: app/src/test/java/org/mifos/mobile/GuarantorListPresenterTest.java
package org.mifos.mobile;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobile.api.DataManager;
import org.mifos.mobile.models.guarantor.GuarantorPayload;
import org.mifos.mobile.presenters.GuarantorListPresenter;
import org.mifos.mobile.ui.views.GuarantorListView;
import org.mifos.mobile.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.IOException;
import java.util.List;

import io.reactivex.Observable;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GuarantorListPresenterTest {
    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    GuarantorListView view;

    private GuarantorListPresenter presenter;
    List<GuarantorPayload> guarantorPayloadList;


    @Before
    public void setUp() throws Exception {
        presenter = new GuarantorListPresenter(context, dataManager);
        presenter.attachView(view);
        guarantorPayloadList = FakeRemoteDataSource.getGuarantorsList();
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testGetGuarantorList() throws IOException {
        long loanId = 1;
        when(dataManager.getGuarantorList(loanId))
                .thenReturn(Observable.just(guarantorPayloadList));
        presenter.getGuarantorList(loanId);
        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showGuarantorListSuccessfully(guarantorPayloadList);
    }

    @Test
    public void testGetGuarantorListOnError() throws IOException {
        long loanId = 1;
        Exception exception = new Exception("ExceptionMessage");
        when(dataManager.getGuarantorList(loanId))
                .thenReturn(Observable.<List<GuarantorPayload>>error(exception));
        presenter.getGuarantorList(loanId);
        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showError(exception.getMessage());
    }
}

File: app/src/test/java/org/mifos/mobile/GuarantorListPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.guarantor.GuarantorPayload
import org.mifos.mobile.presenters.GuarantorListPresenter
import org.mifos.mobile.ui.views.GuarantorListView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

import java.io.IOException

@RunWith(MockitoJUnitRunner::class)
class GuarantorListPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: GuarantorListView? = null
    private var presenter: GuarantorListPresenter? = null
    private var guarantorPayloadList: List<GuarantorPayload?>? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = GuarantorListPresenter(context!!, dataManager!!)
        presenter?.attachView(view)
        guarantorPayloadList = FakeRemoteDataSource.guarantorsList
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(IOException::class)
    fun testGetGuarantorList() {
        val loanId: Long = 1
        Mockito.`when`(dataManager?.getGuarantorList(loanId))
                .thenReturn(Observable.just(guarantorPayloadList))
        presenter?.getGuarantorList(loanId)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showGuarantorListSuccessfully(guarantorPayloadList)
    }

    @Test
    @Throws(IOException::class)
    fun testGetGuarantorListOnError() {
        val loanId: Long = 1
        val exception = Exception("ExceptionMessage")
        Mockito.`when`(dataManager?.getGuarantorList(loanId))
                .thenReturn(Observable.error(exception))
        presenter?.getGuarantorList(loanId)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showError(exception.message)
    }
}